cel/pkg/reconciler/cel/celrun.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2021 The Tekton Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cel
+
+import (
+	"context"
+	"fmt"
+	"github.com/google/cel-go/cel"
+	"github.com/google/cel-go/common/types"
+	"github.com/tektoncd/pipeline/pkg/client/injection/reconciler/pipeline/v1alpha1/run"
+	"knative.dev/pkg/logging"
+
+	v1alpha1 "github.com/tektoncd/pipeline/pkg/apis/pipeline/v1alpha1"
+	v1 "k8s.io/api/core/v1"
+	reconciler "knative.dev/pkg/reconciler"
+)
+
+// newReconciledNormal makes a new reconciler event with event type Normal, and reason RunReconciled.
+func newReconciledNormal(namespace, name string) reconciler.Event {
+	return reconciler.NewEvent(v1.EventTypeNormal, "RunReconciled", "Run reconciled: \"%s/%s\"", namespace, name)
+}
+
+// Reconciler implements controller.Reconciler for Run resources.
+type Reconciler struct {
+}
+
+// Check that our Reconciler implements Interface
+var _ run.Interface = (*Reconciler)(nil)
+
+// ReconcileKind implements Interface.ReconcileKind.
+func (r *Reconciler) ReconcileKind(ctx context.Context, o *v1alpha1.Run) reconciler.Event {
+	logger := logging.FromContext(ctx)
+	logger.Infof("Reconciling Run %s/%s", o.Namespace, o.Name)
+
+	// If the Run has not started, initialize the Condition and set the start time.
+	if !o.HasStarted() {
+		logger.Infof("Starting new Run %s/%s", o.Namespace, o.Name)
+		o.Status.InitializeConditions()
+		// In case node time was not synchronized, when controller has been scheduled to other nodes.
+		if o.Status.StartTime.Sub(o.CreationTimestamp.Time) < 0 {
+			logger.Warnf("Run %s/%s createTimestamp %s is after the Run started %s", o.Namespace, o.Name, o.CreationTimestamp, o.Status.StartTime)
+			o.Status.StartTime = &o.CreationTimestamp
+		}
+	}
+
+	if o.IsDone() {
+		logger.Infof("Run %s/%s is done", o.Namespace, o.Name)
+		return nil
+	}
+
+	// Create a program environment configured with the standard library of CEL functions and macros
+	env, err := cel.NewEnv(cel.Declarations())
```

Comment:
I think we should have all the cel code in a separate file. I would have a struct with methods that perform the different CEL steps. Would be easier to extend if we decide to add custom functions for example... I think :grin:

Reply:
For this custom task, we don't have a CEL CRD type because parameters can be used to provide all that we need (CEL expressions) -- we can define a CEL CRD type in the future if there are other use cases we want to solve for, but would prefer to keep it simple for now